lightllm/utils/max_token_num_utils.py

```diff
@@ -0,0 +1,59 @@
+import torch
+torch.multiprocessing.set_start_method('spawn', force=True) # Fork start method will cause CUDA re-initialization error
+import os
+import json
+
+def get_total_free_gpu_memory(tp):
+    """
+    Returns the total amount of free memory available on all GPUs, in Gigabytes.
+    """
+    devices = min(tp, torch.cuda.device_count())
+    total_free = 0
+    for i in range(devices):
+        total_free += torch.cuda.mem_get_info(i)[0]
+    total_free = total_free / (1024 ** 3)
+    return total_free
+
+def get_total_weight_size(weight_dir):
+    """
+    Returns the total size of all parameters in the model, in Gigabytes.
+    """
+    total_size = 0
+    files = os.listdir(weight_dir)
+    candidate_files = list(filter(lambda x : x.endswith('.safetensors'), files))
+    if len(candidate_files) == 0:
+        candidate_files = list(filter(lambda x : x.endswith('.bin'), files))
+    assert len(candidate_files) != 0, "can only support pytorch tensor and safetensors format for weights."
+    for file in candidate_files:
+        total_size += os.path.getsize(os.path.join(weight_dir, file))
+    total_size = total_size / (1024 ** 3)
+    return total_size
+
+def get_kv_cache_size(model_dir):
```

Comment:
"get_kv_cache_size and xxxx" is best implemented as a member function of TpPartBaseModel and should be inherited and implemented by its subclasses.

Reply:
It seems that `max_total_token_num` (and `batch_max_tokens` that depends on it) gets passed to a lot of subsystems before the model is initialized. We need this value to be ready early.

Is there any way to achieve this if implemented as a member function of TpPartBaseModel.

Reply:
@singularity-s0  You can try to add a method in TpPartBaseModel,  but it is not easy to get and set batch_max_tokens in TpPartBaseModel. Let me think about how to implement it elegantly. What are your suggestions?

Reply:
Ideally, since each instance of LightLLM server is bound to only one model, model configuration can (and should) be loaded before all other subsystems are initialized (because other subsystems may depend on model configuration, as in the case of `max_total_token_num`). A refactor would be the most elegant way to address this.

Other parameters like `max_req_total_len` and `dtype` (which is currently hardcoded to `fp16`) might also be dependent on model `config.json` and would benefit from this refactor.

However I imagine such a refactor would not be easy. Hacky solutions are also available but it is ultimately up to you to decide which way is the best.

Reply:
@singularity-s0 You can write a standalone recommendation program to generate a value for max_total_token_num. that will be more appropriate。